I9 logging config

lowellmower opened the pull request:
### What I Did:
Added the functionality to create a new configuration store which logs upon writes and reads. One can opt for the default log which is the `io.Writer`,`os.Stdout` in a `pkg/log/#Logger` or one can construct their own logging mechanism and initialize the configuration store with such.

### How I Did it:
Introduce LoggingInterface and Wrapper Struct:
```go
// logger.go
type LoggingInterface interface {
	Print(...interface{})
}

type DefaultLogger struct {
	log    LoggingInterface
	suffix string
}
```
Added 'nil-able' Pointer Field on DefaultConfigStore:
```go
// store.go
type DefaultConfigStore struct {
	config ConfigLevelMap
	usedLevels *ConfigLevelHeap
	resolvers map[ConfigLevel]Resolver
	aliases map[string]string
	log *DefaultLogger
}
```
Added Unexported DefaultConfigStore Constructors:
```go
// store.go
func newLogableConfigStoreWith(l LoggingInterface) *DefaultConfigStore
```
Added Exported Venom Constructors:
```go
// venom.go
func NewLogable() *Venom
func NewLogableWith(l LoggingInterface) *Venom
```

### How I Know it Works
Added unit tests and maintained existing tests
```
venom [I9-logging-config] :> make test
go test -coverprofile=coverage.out
flag provided but not defined: -verbose
Usage of test:
PASS
coverage: 96.8% of statements
ok  	github.com/moogar0880/venom	0.047s
```

lowellmower committed:
This commit adds a LoggingInterface to be used by the LoggableConfigStore

An interface by which all logging types must adhere is introduced as well
as a basic struct, DefaultLogger which has basic getter and setter methods
for configuring the LoggingInterface and the structured log line produced.

lowellmower committed:
This commit adds the functionality of the LoggableConfigStore

Functionally, this commit introduces the LoggableConfigStore and the operations
necessary for logging on configuration reads and writes. The additional code
introduced outside of the functionality is for satisfying the ConfigStore
interface for the other types, e.g. DefaultConfigStore and SafeConfigStore.

lowellmower committed:
This commit add the Venom function for getting a LoggableConfigStore

This commit introduces the function by which users will instantiate a new
LoggableConfigStore. The LoggableConfigStore will have all fields set as
defaults and log to STDOUT. These fields can be configured through some
getter and setter methods provided on the DefaultLogger struct.

lowellmower committed:
Commit adds the LogableConfigStore to the store tests

This commit introduces the LogableConfigStore to the store tests but
does not introduce any new tests or test specific to the new config
store type

lowellmower committed:
Commit satisfies the ConfigStore interface for Venom

This commit introduces four functions to Venom, added to the ConfigStore
interface to satisfy the LogableConfigStore requirements. These functions
are:
    SetLogger(LoggingInterface)
    GetLogger() LoggingInterface
    SetPrefix(string)
    SetSuffix(string)

lowellmower committed:
add some basic tests and default no-op logger

moogar0880 commented on global.go:
```diff
@@ -35,7 +35,7 @@ func SetOverride(key string, value interface{}) {
 }
 
-// Get retrievs the requested key from the global venom instance
+// Get retrieves the requested key from the global venom instance
```
okay, _ray_ 🙄

moogar0880 commented on logger.go:
```diff
@@ -0,0 +1,63 @@
+// DefaultLogger is the struct associated with the LoggableConfigStore. It has
+// three configurable fields which ultimately compose structured log lines.
+type DefaultLogger struct {
+    log    LoggingInterface
```
I think we can replace this custom `LoggingInterface` with a generic `io.Writer` since your `DefaultLogger` already almost implements it anyway with that `Write` method. Then your default loggers `log` can just be set to `os.Stdout` instead of the logger that wraps it. Or maybe even the logger can stay, I'd have to check and see if it implements `io.Writer`

lowellmower replied:
Yeah, this was mainly trying to adapt to easily setting the log to something else like:
```go
    ven := venom.NewLogable()
    var log = logrus.New()
    ven.SetLogger(log)
```
But I'll poke a bit at this and see if it we can't just use the `io.Writer` interface

moogar0880 replied:
So I think this is more where the constructors would come in where you could do something like

```go
lcs := venom.NewLogableConfigStoreWith(logrus.New(), venom.NewDefaultConfigStore())
ven := venom.NewWithStore(lcs)
```

So now your `LogableConfigStore` is set to write using a logrus logger and is backed by an actual default config store (or whatever kind of config store), without the `venom.Venom` instance or the underlying `ConfigStore` needing to know or care about what the `LogableConfigStore` is actually doing

lowellmower replied:
Ah I see. Ok, I was trying to keep it to one constructor.

moogar0880 commented on store.go:
```diff
@@ -19,6 +19,10 @@ type ConfigStore interface {
 	Clear()
 	Debug() string
 	Size() int
+	SetLogger(LoggingInterface)
```
I don't think these methods feel like they belong to the `ConfigStore` interface. Which I suspect is what you meant when you said "I’m not sure logging on reads/writes is worth the additional code that would be adding". 

The general idea here was that we wouldn't need to add methods like this to the `ConfigStore` interface, but rather that the `LogableConfigStore` would intercept incoming read/writes and log before proxying the actual read/write down to the wrapped `ConfigStore`. That way only the `LogableConfigStore` needs to know or care about logging. If that makes sense?

lowellmower replied:
I think I understand but it becomes difficult when you then need/want methods which would be unique to the `LogableConfigStore` like getters and setters for the underlying logger and the `Venom` struct being tied to implementing the `Store` like:
```go
type Venom struct {
	Store ConfigStore
}
```

moogar0880 replied:
You do actually still have access to those type specific methods though. When you create the `LogableConfigStore` you can do Things™️ that are `LogableConfigStore` specific before creating the venom instance. Something like,

```go
lcs := venom.NewLogableConfigStore()
// ...do Things™️ to lcs while you know it's a `LogableConfigStore`
ven := venom.NewWithDataStore(lcs)
```

moogar0880 commented on store.go:
```diff
@@ -240,3 +265,120 @@ func (s *SafeConfigStore) Size() int {
+// LogableConfigStore implements the ConfigStore interface and provides a store
+// that is safe to read and write from multiple goroutines.
+type LogableConfigStore struct {
+    c  *DefaultConfigStore
```
we should make this `c ConfigStore` so that it works to help debug any `ConfigStore`

moogar0880 commented on store.go:
```diff
@@ -240,3 +265,120 @@ func (s *SafeConfigStore) Size() int {
+
+// NewLogableConfigStore returns a new LogableConfigStore.
+func NewLogableConfigStore() ConfigStore {
+    return &LogableConfigStore{
```
we should add a new function `NewLogableConfigStoreWith(w io.Writer, c ConfigStore)` to make integration easier. `NewLogableConfigStore` can just pass `NewDefaultLogger()` and `NewDefaultConfigStore()` through as values to that.

moogar0880 commented on store.go:
```diff
@@ -240,3 +265,120 @@ func (s *SafeConfigStore) Size() int {
+func (l *LogableConfigStore) Find(key string) (interface{}, bool) {
+	a, b := l.c.Find(key)
+	if b {
+		l.logger.Write(a)
```
I think that, rather than logging the returned value, it might make more sense to have something like:

```go
val, ok := l.c.Find(key)
l.logger.Write(fmt.Sprintf("find %s:  %v", key, val))
return val, ok
```

That way we'll log that the read occurred even if the value wasn't present in the ConfigStore

lowellmower committed:
Commit removes getters and setters and converts spaces to tabs

This commit introduces changes which largely are centered around the conversion
of spaces to tabs and whitespace fixes noted in code review. This commit also
simplifies the DefaultLogger to be a wrapper around the LoggingInterface and to
house the log suffix sting only.

lowellmower committed:
Commit introduces unit tests for logger.go

This commit introduces unit test cases for the logger.go file which encompasses
the LoggingInterface and DefaultLogger struct.

lowellmower committed:
Commit introduces a test helper function for redirecting os.Stdout

This commit introduces changes related to redirecting stdout in an effort to
thoroughly test the default logger.

lowellmower committed:
Commit introduces new constructor setup for Venom type

This commit introduces changes around how a Venom instance capable of logging
is initialized. There are two constructors, NewLogable() and NewLogableWith()
where the former results in a ConfigStore which defaults to os.Stdout as its
logging mechanism and the latter can take any logging mechanism which
implements the LoggingInterface which is as simple as having a Print() function

lowellmower committed:
Commit introduces functions for creation of LogableConfigStore

This commit introduces changes related to getting a default and a configured
logable configuration store.

lowellmower committed:
Commit simplifies LogableConfigStore test cases for store.go

lowellmower committed:
This commit introduces code review notes from @moogar0880

This commit move the logging field out of the DeafultConfigStore and into the
LoggableConfigStore as well as some general clean up and rewording.

lowellmower committed:
Commit consolidates constructors for loggable config

lowellmower committed:
Commit remove the wrapping struct DefaultLogger

This commit removes the wrapping struct around the Logging interface
which was DefaultLogger. The previously exported method of the struct
responsible for bot log line formatting and logging has been moved to
the package level and now returns a formated string for use by the
Logging interface.

lowellmower committed:
Add LogWrapper interface for structured and configured logs

lowellmower committed:
Remove nested interface and rename things

lowellmower committed:
Commit update usage instructions for logging store
